File: apple/QuenderinKit/Sources/QuenderinKit/DownloadSegments.swift
import Foundation

/// Splits a download into disjoint byte ranges for parallel fetching, and maps per-segment
/// progress back to whole-file progress. Pure (no networking, no files) so it unit-tests in
/// milliseconds — the risky part of a parallel downloader is the arithmetic, not the sockets.
///
/// Why parallel at all: a single TCP stream is capped by `RTT × window`. Against a distant CDN
/// (e.g. HuggingFace's Xet bridge on a US edge from Asia) one stream uses roughly half the link;
/// N concurrent `Range:` requests use the whole pipe. See `docs/` on the download path.
public struct DownloadSegments: Sendable, Equatable {
    public struct Segment: Sendable, Equatable {
        public let index: Int
        /// Inclusive first byte.
        public let start: Int64
        /// Inclusive last byte.
        public let end: Int64
        public var length: Int64 { end - start + 1 }

        public init(index: Int, start: Int64, end: Int64) {
            self.index = index
            self.start = start
            self.end = end
        }
    }

    public let total: Int64
    public let segments: [Segment]

    /// Even split with the remainder spread over the first segments, so the union is exactly
    /// `[0, total)` with no gaps or overlaps. Never more segments than bytes.
    public init(totalBytes: Int64, segmentCount: Int) {
        precondition(totalBytes > 0, "segment plan needs a known, positive size")
        let n = Int(Swift.max(1, Swift.min(Int64(segmentCount), totalBytes)))
        let base = totalBytes / Int64(n)
        let remainder = totalBytes % Int64(n)
        var segs: [Segment] = []
        var start: Int64 = 0
        for i in 0..<n {
            let len = base + (Int64(i) < remainder ? 1 : 0)
            segs.append(Segment(index: i, start: start, end: start + len - 1))
            start += len
        }
        self.total = totalBytes
        self.segments = segs
    }

    /// How many parallel connections to open. Scales with size (~1 per 32 MB) so a small model
    /// isn't fanned out pointlessly, and is deliberately lower on cellular (money + battery).
    public static func connectionCount(totalBytes: Int64, isCellular: Bool, max: Int = 6) -> Int {
        guard totalBytes > 0 else { return 1 }
        let perSegment: Int64 = 32 << 20
        let bySize = Int((totalBytes + perSegment - 1) / perSegment)
        let ceiling = isCellular ? Swift.min(3, max) : max
        return Swift.max(1, Swift.min(ceiling, bySize))
    }

    /// The whole-file fraction (0...1) for a per-segment downloaded-byte vector. Extra bytes beyond
    /// a segment's length are ignored, so a mis-sized response can never push progress past 1.0.
    public func fraction(downloadedPerSegment: [Int64]) -> Double {
        guard total > 0 else { return 0 }
        var done: Int64 = 0
        for (i, seg) in segments.enumerated() {
            let have = i < downloadedPerSegment.count ? downloadedPerSegment[i] : 0
            done += Swift.min(seg.length, Swift.max(0, have))
        }
        return Double(done) / Double(total)
    }
}

File: android/quenderin-core/src/main/kotlin/ai/quenderin/core/DownloadSegments.kt
package ai.quenderin.core

/**
 * Pure segment arithmetic for the parallel download path — the Kotlin twin of Swift
 * `DownloadSegments`. Splits a known-size file into disjoint `Range:` segments and maps
 * per-segment progress back to a whole-file fraction. No I/O, no threads, so it verifies
 * in the headless harness.
 *
 * Why parallel: one TCP stream is capped by `RTT x window`, so a single connection to a
 * distant CDN uses roughly half the link; N concurrent range requests use the pipe.
 */
object DownloadSegments {

    data class Segment(val index: Int, val start: Long, val end: Long) {
        /** Inclusive length in bytes. */
        val length: Long get() = end - start + 1
    }

    data class Plan(val total: Long, val segments: List<Segment>)

    /** Even split with the remainder on the first segments — the union is exactly `[0, total)`. */
    fun plan(totalBytes: Long, segmentCount: Int): Plan {
        require(totalBytes > 0) { "segment plan needs a known, positive size" }
        val n = maxOf(1, minOf(segmentCount.toLong(), totalBytes)).toInt()
        val base = totalBytes / n
        val remainder = totalBytes % n
        val segments = ArrayList<Segment>(n)
        var start = 0L
        for (i in 0 until n) {
            val len = base + if (i < remainder) 1 else 0
            segments.add(Segment(i, start, start + len - 1))
            start += len
        }
        return Plan(totalBytes, segments)
    }

    /**
     * How many parallel connections to open. Scales with size (~1 per 32 MB) so a small model
     * isn't fanned out pointlessly, and is lower on cellular (money + battery).
     */
    fun connectionCount(totalBytes: Long, isCellular: Boolean, max: Int = 6): Int {
        if (totalBytes <= 0) return 1
        val perSegment = 32L shl 20
        val bySize = ((totalBytes + perSegment - 1) / perSegment).toInt()
        val ceiling = if (isCellular) minOf(3, max) else max
        return maxOf(1, minOf(ceiling, bySize))
    }

    /** Whole-file fraction (0...1); extra bytes beyond a segment are ignored, so never > 1.0. */
    fun fraction(plan: Plan, downloadedPerSegment: LongArray): Double {
        if (plan.total <= 0) return 0.0
        var done = 0L
        plan.segments.forEachIndexed { i, seg ->
            val have = if (i < downloadedPerSegment.size) downloadedPerSegment[i] else 0L
            done += minOf(seg.length, maxOf(0L, have))
        }
        return done.toDouble() / plan.total
    }
}
